File: projecteuler/python/problem_4.py
'''
Created on 03/09/2011

@author: Breno Martinusso

PRoject Euler - Problem 4

A palindromic number reads the same both ways. 
The largest palindrome made from the product of two 2-digit numbers is 9009 = 91 x 99.

Find the largest palindrome made from the product of two 3-digit numbers.

Answer: 906609
'''

def is_palindrome(number):
    s = str(number)
    return s == s[::-1]

largest = 0
for n1 in xrange(999, 100, -1):
    for n2 in xrange(n1, 100, -1):
        calc = n1 * n2
        if is_palindrome(calc) and (calc > largest):
            largest = calc
print largest

File: projecteuler/go/problem4.go
package projecteuler

import "strconv"

// Problem 4
// Largest palindrome product
// A palindromic number reads the same both ways. The largest palindrome made from the product of two 2-digit numbers is 9009 = 91 × 99.
// Find the largest palindrome made from the product of two 3-digit numbers.

const (
	answerProblem4 = 906609

	max3DigitNumber = 999
	min3DigitNumber = 100
)

func problem4() int {
	largestPalindrome := 0
	for i := max3DigitNumber; i >= min3DigitNumber; i-- {
		for j := i; j >= min3DigitNumber; j-- {
			product := i * j

			if product < largestPalindrome {
				break
			}

			if isPalindromic(product) {
				largestPalindrome = product
			}
		}
	}
	return largestPalindrome
}

func isPalindromic(number int) bool {
	n := strconv.Itoa(number)
	for i := 0; i < len(n)/2; i++ {
		if n[i] != n[len(n)-i-1] {
			return false
		}
	}
	return true
}
